File: frontend/src/api/roulette.ts
import { apiRequest } from './client';

export interface RouletteBet {
  bet_type: string;
  bet_value: string;
  amount: number;
}

export interface RouletteBetResult {
  bet_type: string;
  bet_value: string;
  amount: number;
  payout: number;
}

export interface RouletteSpinResponse {
  number: number;
  color: string;
  results: RouletteBetResult[];
  total_bet: number;
  total_win: number;
  new_balance: number;
}

export interface RouletteHistory {
  id: number;
  user_id: number;
  bet_amount: number;
  win_amount: number;
  bet_data: RouletteBetResult[];
  result_number: number;
  result_color: string;
  created_at: string;
}

export interface RouletteStats {
  total_spins: number;
  total_bet: number;
  total_won: number;
  net_result: number;
  biggest_win: number;
}

export interface RouletteTopWinner {
  user_id: number;
  user_name: string;
  total_spins: number;
  total_bet: number;
  total_won: number;
  net_result: number;
}

export interface RouletteAdminStats {
  total_spins: number;
  total_bet: number;
  total_won: number;
  house_profit: number;
  actual_rtp: number;
  theoretical_rtp: number;
  top_winners: RouletteTopWinner[];
  period_start: string | null;
  period_end: string | null;
  enabled: boolean;
}

export interface RouletteStatusResponse {
  enabled: boolean;
}

export function spinRoulette(bets: RouletteBet[]): Promise<RouletteSpinResponse> {
  return apiRequest<RouletteSpinResponse>('/roulette/spin', {
    method: 'POST',
    body: JSON.stringify({ bets }),
  });
}

export function getRouletteHistory(limit: number = 50): Promise<RouletteHistory[]> {
  return apiRequest<RouletteHistory[]>(`/roulette/history?limit=${limit}`);
}

export function getRouletteStats(): Promise<RouletteStats> {
  return apiRequest<RouletteStats>('/roulette/stats');
}

export function getRouletteAdminStats(scope: 'fiscal_period' | 'all_time' = 'fiscal_period'): Promise<RouletteAdminStats> {
  return apiRequest<RouletteAdminStats>(`/roulette/admin/stats?scope=${scope}`);
}

export function getRouletteStatus(): Promise<RouletteStatusResponse> {
  return apiRequest<RouletteStatusResponse>('/roulette/status');
}

export function toggleRoulette(enabled: boolean): Promise<RouletteStatusResponse> {
  return apiRequest<RouletteStatusResponse>('/roulette/admin/toggle', {
    method: 'PUT',
    body: JSON.stringify({ enabled }),
  });
}

File: backend/app/schemas/roulette.py
from datetime import datetime
from pydantic import BaseModel


class RouletteBet(BaseModel):
    bet_type: str
    bet_value: str
    amount: float


class RouletteSpinRequest(BaseModel):
    bets: list[RouletteBet]


class RouletteBetResult(BaseModel):
    bet_type: str
    bet_value: str
    amount: float
    payout: float


class RouletteSpinResponse(BaseModel):
    number: int
    color: str
    results: list[RouletteBetResult]
    total_bet: float
    total_win: float
    new_balance: float


class RouletteHistory(BaseModel):
    id: int
    user_id: int
    bet_amount: float
    win_amount: float
    bet_data: list[RouletteBetResult]
    result_number: int
    result_color: str
    created_at: datetime

    class Config:
        from_attributes = True


class RouletteStats(BaseModel):
    total_spins: int
    total_bet: float
    total_won: float
    net_result: float
    biggest_win: float


class RouletteTopWinner(BaseModel):
    user_id: int
    user_name: str
    total_spins: int
    total_bet: float
    total_won: float
    net_result: float


class RouletteAdminStats(BaseModel):
    total_spins: int
    total_bet: float
    total_won: float
    house_profit: float
    actual_rtp: float
    theoretical_rtp: float
    top_winners: list[RouletteTopWinner]
    period_start: datetime | None
    period_end: datetime | None
    enabled: bool


class RouletteStatusResponse(BaseModel):
    enabled: bool


class RouletteToggleRequest(BaseModel):
    enabled: bool
